### ISSUE TITLE
[10.0] PoS recovery mechanism unable to create recovery session

### ISSUE THREAD

**Issue by aisopuro** [odoo/odoo#15101]
#### Impacted versions:
10.0 confirmed, though it may be present in earlier versions (particularly 9.0 and/or saas)
#### Steps to reproduce:
- Create some orders in a PoS session, but without an internet connection so they are not sent to the server
- Close the frontend UI with the orders still not being sent
  - You will be shown a warning, but close the frontend anyway
- Close the session in the backend
- Open a new session in the same PoS, with the same user

#### Current behavior:
- As soon as the PoS frontend is loaded, the user shown an error popup with a traceback, warning them that they cannot create multiple sessions for the same point of sale. The tracebcak is at the end of this issue description.
- No new orders appear in the backend, and manual synchronization does not work, even if the internet connection is working. The connections status simply shows a red warning triangle, and trying to manually sync orders by pressing it does nothing.
#### Expected behavior:
- Whether or not the user is warned, they should be able to accept new orders in the current session, and be able to handle the problematic orders later
#### Code details:

The problem appears to stem from the fact that the recovery mechanism for PoS orders has undergone a lot of complicated rework in the past two years, and it may be that some of that work has gotten incorrectly merged. A rough timeline:
1. bed409322d71935903f6c2e9880bc6462c58b78e @mart-e adds the `rescue` flag on the `pos.session` model, allowing such sessions to circumvent the constraint checking (so that the same user and same points of sale can have multiple recovery sessions open simultaneously)
1. 23d7caa8d50de8b5307936af2750c2ffe9c82540 @jorenvo reverts that functionality in order to fix something
1. e892c574825c19598fe3c243fcb651f8b6d50d0f @odony adds back the recovery functionality.

The commit by @odony seems to have been done using the old API, is it maybe meant for a saas branch originally? It seems possible that that code was not meant to get merged into 10.0. It was merged into 9.0 in this pull request: https://github.com/brain-tec/odoo/pull/111

The problem is that @odony's code requires the existence of the `recovery` field on `pos.session`, but as of e91a8d428ff0ca953a7ed2967d3be3f518f05b49, the field is not present in the model, apparently due to the revert done by @jorenvo.

Perhaps the commit authors are better equipped to comment on what would be the best way to fix this: should the recovery mechanism work like @mart-e intended, or was it removed with good reason?

Here is the traceback that is shown to the user when they open the PoS frontend with a new session:
```
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/odoo/http.py", line 638, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/usr/lib/python2.7/dist-packages/odoo/http.py", line 675, in dispatch
    result = self._call_function(**self.params)
  File "/usr/lib/python2.7/dist-packages/odoo/http.py", line 331, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/usr/lib/python2.7/dist-packages/odoo/service/model.py", line 119, in wrapper
    return f(dbname, *args, **kwargs)
  File "/usr/lib/python2.7/dist-packages/odoo/http.py", line 324, in checked_call
    result = self.endpoint(*a, **kw)
  File "/usr/lib/python2.7/dist-packages/odoo/http.py", line 933, in __call__
    return self.method(*args, **kw)
  File "/usr/lib/python2.7/dist-packages/odoo/http.py", line 504, in response_wrap
    response = f(*args, **kw)
  File "/usr/lib/python2.7/dist-packages/odoo/addons/web/controllers/main.py", line 862, in call_kw
    return self._call_kw(model, method, args, kwargs)
  File "/usr/lib/python2.7/dist-packages/odoo/addons/web/controllers/main.py", line 854, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/usr/lib/python2.7/dist-packages/odoo/api.py", line 679, in call_kw
    return call_kw_model(method, model, args, kwargs)
  File "/usr/lib/python2.7/dist-packages/odoo/api.py", line 664, in call_kw_model
    result = method(recs, *args, **kwargs)
  File "/opt/odoo_addons/pos_orders/models/wk_orders.py", line 16, in create_from_ui
    order_ids = super(PosOrder,self).create_from_ui(orders);
  File "/usr/lib/python2.7/dist-packages/odoo/addons/point_of_sale/models/pos_order.py", line 503, in create_from_ui
    pos_order = self._process_order(order)
  File "/usr/lib/python2.7/dist-packages/odoo/addons/point_of_sale/models/pos_order.py", line 102, in _process_order
    pos_order['pos_session_id'] = self._get_valid_session(pos_order).id
  File "/usr/lib/python2.7/dist-packages/odoo/addons/point_of_sale/models/pos_order.py", line 73, in _get_valid_session
    'rescue': True,    # avoid conflict with live sessions
  File "/usr/lib/python2.7/dist-packages/odoo/addons/point_of_sale/models/pos_session.py", line 193, in create
    res = super(PosSession, self.with_context(ctx).sudo(uid)).create(values)
  File "/usr/lib/python2.7/dist-packages/odoo/models.py", line 3810, in create
    record = self.browse(self._create(old_vals))
  File "/usr/lib/python2.7/dist-packages/odoo/models.py", line 3968, in _create
    self._validate_fields(vals)
  File "/usr/lib/python2.7/dist-packages/odoo/models.py", line 1069, in _validate_fields
    check(self)
  File "/usr/lib/python2.7/dist-packages/odoo/addons/point_of_sale/models/pos_session.py", line 134, in _check_pos_config
    raise ValidationError(_("You cannot create two active sessions related to the same point of sale!"))
ValidationError: (u'You cannot create two active sessions related to the same point of sale!', None)
```

**Comment by Yenthe666** [comment]
Hi @aisopuro,

Thank you for your report.
I can confirm it is a bug and @odony will handle this.
We'll give you feedback as soon as possible. For now I'd like to thank you for the detailed description and work you've put in this report, it is one of the best reports I've ever seen here. Thanks 👍 

Regards,
Yenthe

**Comment by odony** [comment]
Thanks for the nice summary! To further clarify the rationale/timeline:

1. bed409322d71935903f6c2e9880bc6462c58b78e @mart-e adds the first recovery mechanism, based on a new `rescue` field on the pos.session, to fix a shortcoming of the POS system.  Due to the new field, this is done in the then-`master`, aka future 9.0.
1. 35e9fa5c410db78fbad9492afc57b5e7c05a9cb4 Later, @jorenvo is asked to fix the same problem in 8.0, and imagines a fix without the flag (can't change schema in stable), instead reusing a newer session if one exists.
1. 87a3fe92de43dd7fce7e3364a5ca8dbb417ca7f8 Seeing that we now had two independent ways to fix the same problem (8.0 vs 9.0), @jorenvo then proceeds to copy his newer solution to 9.0, making the `rescue` flag unused.
1. 23d7caa8d50de8b5307936af2750c2ffe9c82540 @jorenvo copies his solution to the new `master` (future 10.0) and removes the `rescue` flag, now unused.
1. e892c574825c19598fe3c243fcb651f8b6d50d0f based on customer feedback @odony identifies several problems with the recovery system introduced by @jorenvo, stemming from the lack of way to separate "recovered orders" from normal orders (required for users, in order to save/control them appropriately), and a brittle way to create recovery sessions, failing most of the time. For 9.0 the `rescue` flag is still present so the initial recovery mechanism can be restored, and it can be even improved (better names for sessions, for easier identification).
1. e91a8d428ff0ca953a7ed2967d3be3f518f05b49 @KangOl attempts to forward the patch from e892c574825c19598fe3c243fcb651f8b6d50d0f up to 10.0, but this cannot work reliably due to the lack of way to distinguish *rescue* sessions from normal sessions. (This is the root cause of the bug you're seeing)

Fortunately each change is well explained in the respective commit message, so it's (relatively) easy to trace.

After establishing the above, what should we do? For 8.0 and 9.0, I think we should keep the recovery mechanism as they are now. Each was designed for that specific version and has the corresponding db schema.

For 10.0, I'd like to support the same recovery mechanism as in 9.0, otherwise we still have all the problems highlighted in e892c574825c19598fe3c243fcb651f8b6d50d0f's commit message, which could technically be seen as a regression. Unfortunately we do not have the `rescue` flag anymore, so it's difficult.
As a temporary alternative, I'm tempted to suggest using the _name_ of the session as a way to distinguish rescue sessions, and exclude them from the "unique constraint". I know this is not pretty, but users can't normally change session names (derived from a POS sequence automatically), so it's relatively safe, provided we are careful not to translate the keyword.
We would be able to more or less keep the same code as for 9.0, except that each occurrence of the `rescue = True` would become `name ilike [RESCUE]` or something similar.

Then for master/11.0 we would simply need to do the reverse change, and add the `rescue` field back.

@mart-e @jorenvo: I'd like your opinions here before creating another space-time divergence :sweat_smile: I really believe we need an explicit rescue system, otherwise users don't understand what is happening, and they can't react/improve their workflow. Explicit > Implicit ;-) In addition, the `rescue` makes the recovery session creation very robust, with no risk of triggering the unique constraint.

In the meantime we could revert the patch from e91a8d428ff0ca953a7ed2967d3be3f518f05b49 to fix the immediate problem.

**Comment by jorenvo** [comment]
@odony thanks for the detailed timeline. As you correctly deduced I initially forward-ported what I did in 8.0 because it was a bit simpler and because I wanted to have one approach across all the versions.

At the time I did not see the need for a dedicated rescue session (I think I talked about it briefly with @mart-e ). I do now though and agree with the reasons you outlined in your commit message. So I think it's a good idea to reimplement rescue sessions.

> For 8.0 and 9.0, I think we should keep the recovery mechanism as they are now. Each was designed for that specific version and has the corresponding db schema.

I agree.

> For 10.0, I'd like to support the same recovery mechanism as in 9.0, otherwise we still have all the problems highlighted in e892c57's commit message, which could technically be seen as a regression.

Your suggested approach would be relatively painless to maintain/port since the logic remains pretty much the same as in 9.0. So I think it's a good idea. This way we avoid having multiple approaches and we avoid the issues mentioned in your commit message.

> Then for master/11.0 we would simply need to do the reverse change, and add the rescue field back.

Sounds good.
